File: src/lexer.cpp
// lexer.cpp

#include "../include/lexer.h"

#include <iostream>

Lexer::Lexer(std::string src) : source(std::move(src)), position(0), line(1), column(1) {}

const static std::unordered_map<std::string, TokenType> keywords = {
    {"fn", TokenType::FUNCTION},
    {"return", TokenType::RETURN},
    {"int",TokenType::INT},
    {"char",TokenType::CHAR},
    {"module",TokenType::MODULE},
    {"use",TokenType::USE},
};

void Lexer::SkipWhitespace() {
    while (!IsAtEnd() && isspace(Peek())) {
        Advance();
    }
}

char Lexer::Advance() {
    const char32_t current = source[position++];
    if (current == '\n') {  // 换行符
        line++;
        column = 1;
    } else {
        column++;
    }
    return current;
}

char32_t Lexer::Peek() const {
    return source[position];
}

bool Lexer::IsAtEnd() const {
    return position >= source.length();
}

std::string Lexer::ReadIdentifier() {
    std::string result;
    while (!IsAtEnd() && (isalpha(Peek()) || isdigit(Peek()) || Peek() == '_')) {
        result += Advance();
    }
    return result;
}

std::string Lexer::ReadNumber() {
    std::string result;
    while (!IsAtEnd() && isdigit(Peek())) {
        result += Advance();
    }
    return result;
}

// 判断是否是汉字首字节（1110xxxx）
bool Lexer::IsChineseCharStart(const char32_t c) {
    return (c & 0xE0) == 0xE0; // 0xE0 = 11100000
}

// 吞掉完整UTF-8汉字（3字节）
Token Lexer::ExtractHanzi() {
    std::string zhchar;
    zhchar += source[position++]; // 吞第1字节
    zhchar += source[position++]; // 吞第2字节
    zhchar += source[position++]; // 吞第3字节
    return {TokenType::ZHCHAR, zhchar, line, column};
}

// 判断是否是有效UTF-8起始字节
bool IsUTF8StartByte(const unsigned char c) {
    if ((c & 0x80) == 0) return false;  // ASCII字符交给原有逻辑
    return (c & 0xC0) == 0xC0;          // 110xxxxx或111xxxxx
}

// 吞掉一个完整UTF-8字符（1-4字节）
Token Lexer::ExtractUTF8Char() {
    int len = 1;
    if ((source[position] & 0xF0) == 0xF0) len = 4;  // 11110xxx
    else if ((source[position] & 0xE0) == 0xE0) len = 3; // 1110xxxx
    else if ((source[position] & 0xC0) == 0xC0) len = 2; // 110xxxxx

    std::string utf8_char;
    for (int i = 0; i < len && !IsAtEnd(); ++i) {
        utf8_char += source[position++];
    }
    return {TokenType::UNICODE_CHAR, utf8_char, line, column};
}

Token Lexer::NextToken() {
    // 跳过空格、文件结尾
    SkipWhitespace();

    // 是否为文件结尾
    if (IsAtEnd()) {
        return {TokenType::EOF_TOKEN, "", line, column};
    }

    if (IsUTF8StartByte(source[position])) return ExtractUTF8Char();
    if (IsChineseCharStart(source[position])) return ExtractHanzi();

    char c = Peek();

    // 标识符或关键字
    if (isalpha(c) || c == '_') {
        std::string ident = ReadIdentifier();
        const auto it = keywords.find(ident);
        if (it != keywords.end()) {
            return {it->second, ident, line, column};
        }
        return {TokenType::IDENTIFIER, ident, line, column};
    }

    if (isdigit(c)) {
        std::string ident = ReadIdentifier();
        const auto it = keywords.find(ident);
        if (it != keywords.end()) {
            return {it->second, ident, line, column};
        }
    }

    // 数字
    if (isdigit(c)) {
        std::string number = ReadNumber();
        return {TokenType::NUMBER, number, line, column};
    }

    c = Advance();

    // 单符号识别
    switch (c) {
        case '+': return {TokenType::PLUS, "+", line, column};
        case '-': return {TokenType::MINUS, "-", line, column};
        case '*': return {TokenType::MULTIPLICATION, "*", line, column};
        case '/': return {TokenType::DIVIDE, "/", line, column};
        case '(': return {TokenType::LEFT_PAREN, "(", line, column};
        case ')': return {TokenType::RIGHT_PAREN, ")", line, column};
        case ',': return {TokenType::COMMA, ",", line, column};
        case ';': return {TokenType::SEMICOLON, ";", line, column};
        case ':': {
            if (Peek() == ':') {
                Advance(); return {TokenType::COLON_COLON, "::", line, column};
            }
            return {TokenType::COLON, ":", line, column};
        }
        case '.': return {TokenType::DOT, ".", line, column};
        case '[': return {TokenType::LBRACKET, "[", line, column};
        case ']': return {TokenType::RBRACKET, "]", line, column};
        case '{': return {TokenType::LEFT_BRACE, "{", line, column};
        case '}': return {TokenType::RIGHT_BRACE, "}", line, column};
        case '<': return {TokenType::LESS, "<", line, column};
        case '>': return {TokenType::GREATER, ">", line, column};
        case '%': return {TokenType::PERCENT, "%", line, column};
        case '\\': return {TokenType::BACKSLASH, "\\", line, column};
        case '\'': return {TokenType::SINGLE_QUOTE, "\'", line, column};
        case '\"': return {TokenType::DOUBLE_QUOTE, "\"", line, column};
        case '?': return {TokenType::QUESTION, "?", line, column};
        case '|': return {TokenType::PIPE, "|", line, column};
        case '&': return {TokenType::AND, "&", line, column};
        case '=': return {TokenType::EQUAL, "=", line, column};
        default: printf("[line %d, column %d]\033[31m error:\033[0m Unknown token: %c\n", line, column, c);
            return {TokenType::UNKNOWN, "", line, column};
    }

}

File: src/lexer.rs
pub enum TokenType {
    Unknown,
    Eof,
    Identifier(String),
    Number(String),
    Func,
}

pub struct Token {
    tok: TokenType,
    ident: Vec<char>,
    num_val: f64,
    position: usize,
}

pub struct Lexer {
    cur_char: Token,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Self {
            cur_char: Token {
                tok: TokenType::Unknown,
                ident: input.chars().collect(),
                num_val: 0.0,
                position: 0,
            },
        }
    }

    pub fn next_tok(&mut self) -> TokenType {
        self.skip_whitespace();

        if self.is_eof() {
            return TokenType::Eof;
        }

        let ch = self.current_char();

        match ch {
            'a'..='z' | 'A'..='Z' | '_' => self.read_identifier(),
            '0'..='9' => self.read_number(),

            _ => panic!("Unknown identifier: {}", ch),
        }
    }

    fn read_identifier(&mut self) -> TokenType {
        let start = self.cur_char.position;
        while !self.is_eof() && self.current_char().is_alphanumeric() {
            self.advance();
        }

        let ident: String = self.cur_char.ident[start..self.cur_char.position]
            .iter()
            .collect();

        println!("IdentifierStr: {}", ident);

        match ident.as_str() {
            "func" => TokenType::Func,
            _ => return TokenType::Identifier(ident),
        }
    }

    fn read_number(&mut self) -> TokenType {
        let start = self.cur_char.position;
        while !self.is_eof() && self.current_char().is_digit(10) {
            self.advance();
        }

        if !self.is_eof() && self.current_char() == '.' {
            self.advance();
            while !self.is_eof() && self.current_char().is_digit(10) {
                self.advance();
            }
        }

        let num_str: String = self.cur_char.ident[start..self.cur_char.position]
            .iter()
            .collect();
        self.cur_char.num_val = num_str.parse::<f64>().unwrap();
        println!("NumVal: {}", self.cur_char.num_val);
        TokenType::Number(self.cur_char.num_val.to_string())
    }

    fn current_char(&self) -> char {
        self.cur_char.ident[self.cur_char.position]
    }

    fn advance(&mut self) {
        self.cur_char.position += 1
    }

    fn is_eof(&self) -> bool {
        self.cur_char.position >= self.cur_char.ident.len()
    }

    fn skip_whitespace(&mut self) {
        while !self.is_eof() && self.current_char().is_whitespace() {
            self.advance()
        }
    }
}
